apiserver/root.go

```diff
@@ -143,16 +159,125 @@ func (r *apiHandler) Authorizer() facade.Authorizer {
 	return r
 }
 
-func (r *apiHandler) getRpcConn() *rpc.Conn {
-	return r.rpcConn
+// CloseConn closes the underlying connection.
+func (r *apiHandler) CloseConn() error {
+	return r.rpcConn.Close()
 }
 
 // Kill implements rpc.Killer, cleaning up any resources that need
 // cleaning up to ensure that all outstanding requests return.
 func (r *apiHandler) Kill() {
+	r.watcherRegistry.Kill()
+	if err := r.watcherRegistry.Wait(); err != nil {
+		logger.Infof("error waiting for watcher registry to stop: %v", err)
+	}
 	r.resources.StopAll()
 }
 
+// AuthMachineAgent returns whether the current client is a machine agent.
+// TODO(controlleragent) - add AuthControllerAgent function
+func (r *apiHandler) AuthMachineAgent() bool {
+	_, isMachine := r.GetAuthTag().(names.MachineTag)
+	_, isControllerAgent := r.GetAuthTag().(names.ControllerAgentTag)
+	return isMachine || isControllerAgent
+}
+
+// AuthModelAgent return whether the current client is a model agent
+func (r *apiHandler) AuthModelAgent() bool {
```

Comment:
Seems to me we could something nicer than these with generics.
Not the concern of this patch of course.